cmd/object-handlers.go

```diff
@@ -331,29 +336,51 @@ func (api objectAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Req
 		writeErrorResponse(w, s3Error, r.URL)
 		return
 	}
+
+	getObjectNInfo := objectAPI.GetObjectNInfo
+	if api.CacheAPI() != nil {
+		getObjectNInfo = api.CacheAPI().GetObjectNInfo
+	}
+
+	// Get request range.
+	var rs *HTTPRangeSpec
+	rangeHeader := r.Header.Get("Range")
+	if rangeHeader != "" {
+		var err error
+		if rs, err = parseRequestRangeSpec(rangeHeader); err != nil {
+			// Handle only errInvalidRange. Ignore other
+			// parse error and treat it as regular Get
+			// request like Amazon S3.
+			if err == errInvalidRange {
+				writeErrorResponse(w, ErrInvalidRange, r.URL)
+				return
+			}
+
+			// log the error.
+			logger.LogIf(ctx, err)
+		}
+	}
+
+	objInfo, reader, err := getObjectNInfo(ctx, bucket, object, rs)
```

Comment:
Ah, I missed this in review - but reader and err may be non-nil in some circumstances - so this could leave the reader unclosed (and the lock stuck). That's why the last version of the code had:

```
if reader != nil {
    defer reader.Close()
}
```

Reply:
there is `defer reader.close()` later in the code.